pkg/endpoint/bpf.go

```diff
@@ -1146,3 +1147,63 @@ func (e *Endpoint) RequireRouting() (required bool) {
 func (e *Endpoint) RequireEndpointRoute() bool {
 	return e.DatapathConfiguration.InstallEndpointRoute
 }
+
+type netNSIPVLANConfigFunc func(netNsPath string) (mapFD, mapID int, err error)
+
+// FinishIPVLANInit finishes configuring ipvlan slave device of the given endpoint.
+//
+// Unfortunately, Docker libnetwork itself moves a netdev to netns of a container
+// after the Cilium libnetwork plugin driver has responded to a `JoinEndpoint`
+// request. During the move, the netdev qdisc's get flushed by the kernel. Therefore,
+// we need to configure the ipvlan slave device in two stages.
+//
+// Because the function can be called many times for the same container in parallel,
+// we need to make the function idempotent. This is achieved by checking
+// whether the datapath map has been pinned, which indicates previous
+// successful invocation of the function for the same container, before executing
+// the configuration stages.
+//
+// FIXME: Because of the libnetwork limitation mentioned above, we cannot enforce
+// policies for an ipvlan slave before a process of a container has started. So,
+// this enables a window between the two stages during which ALL container traffic
+// is allowed.
+func (e *Endpoint) FinishIPVLANInit(netNSIPVLANConfigFunc netNSIPVLANConfigFunc, netNsPath string) error {
+	if netNsPath == "" {
+		return fmt.Errorf("netNsPath is empty")
+	}
+
+	// Just ignore if the endpoint is dying
+	if err := e.LockAlive(); err != nil {
+		return nil
+	}
+	defer e.Unlock()
+
+	// No need to finish IPVLAN initialization for Docker if the endpoint isn't
+	// running with Docker.
+	if e.DockerNetworkID == "" {
+		return nil
+	}
+
+	if e.isDatapathMapPinnedLocked() {
+		// The datapath map is pinned which implies that the post-initialization
+		// for the ipvlan slave has been successfully performed
+		return nil
+	}
+
+	mapFD, mapID, err := netNSIPVLANConfigFunc(netNsPath)
```

Comment:
Can't we just call `connector.ConfigureNetNSForIPVLAN()`? Then we wouldn't need to pass the callback to this method.

Reply:
I want to avoid passing in netlink as a transitive dependency into the endpoint package. It makes running unit tests require Linux.

Reply:
> It makes running unit tests require Linux.

Probably we could create `pkg/endpoint/connector/ipvlan_{linux,darwin}.go` or add the build constraint `// +build linux`` in a unit test module.

The problem I see with passing the callback is that we leak the internal implementation detail to a user of the function. Also, it gives a wrong impression to the user that there exists multiple different ways to configure netns and friends for the ipvlan.

Reply:
I see your intention. As long as we have such unit test with failure conditions, the indirection makes sense. However, it comes with a price that for a reader it's more difficult to follow the ipvlan setup which is already quite complicated.

> What are your thoughts on me adding this functionality to the pkg/datapath:Datapath interface?

If the Datapath interface can accommodate a custom ipvlan-related method and an instance of the interface can be accessed from the Endpoint, then yes. Or maybe it's time `type Connector interface`?

Reply:
@brb check the latest commit - I believe it addresses your concerns as well as puts it behind an interface.

Reply:
Amazing!